### PR TITLE
expose health check via fastapi

### PR DESCRIPTION
**Pull request by kosta**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -2,17 +2,17 @@ FROM python:3.10 as base
...
 RUN poetry config virtualenvs.create false
 
+COPY poetry.lock .
+COPY pyproject.toml .
+COPY setup.py .
 RUN poetry install --no-dev --no-interaction --no-ansi
 
-ENTRYPOINT python app.py
+COPY sched_slack_bot sched_slack_bot
+COPY bin/app.py app.py
```
Comment (kosta) [inline Dockerfile:16]: Drive-by: Re-order Dockerfile to not install everything from scratch if source file changes.

File: bin/app.py [hunk 2]
```diff
@@ -2,11 +2,9 @@
...
-if __name__ == "__main__":
-    logging.basicConfig(level=logging.DEBUG)
+logging.basicConfig(level=logging.DEBUG)
 
-    controller = AppController()
+logger = logging.getLogger(__name__)
 
-    controller.start()
+controller = AppController()
+api = controller.start()
```
Comment (kosta) [inline bin/app.py:10]: As this is not started via uvicorn, `__name__ == "__main__"` no longer works.

File: Dockerfile [hunk 3]
```diff
@@ -2,17 +2,17 @@ FROM python:3.10 as base
...
-
 RUN apt-get -y install curl
 RUN curl -sSL https://install.python-poetry.org/install-poetry.py | python -
 ENV PATH=/root/.local/bin:$PATH
 RUN poetry config virtualenvs.create false
 
+COPY poetry.lock .
+COPY pyproject.toml .
+COPY setup.py .
 RUN poetry install --no-dev --no-interaction --no-ansi
```
Comment (Germandrummer92) [inline Dockerfile:13]: you need to --no-root here and then poetry install again after copying. otherwise this first poetry install will fail

File: sched_slack_bot/controller.py [hunk 4]
```diff
@@ -91,7 +93,19 @@ def start(self) -> None:
...
+
+        @api.post("/slack/events")
+        async def endpoint(req: Request) -> Response:
+            return await app_handler.handle(req)
+
+        @api.get("/health")
+        async def health(req: Request) -> Response:
+            return Response(status_code=200)
+
+        return api
```
Comment (kosta) [inline sched_slack_bot/controller.py:108]: Inspired by https://github.com/slackapi/bolt-python/blob/main/examples/fastapi/app.py

File: sched_slack_bot/controller.py [hunk 5]
```diff
@@ -91,7 +93,19 @@ def start(self) -> None:
 
         self._start_all_saved_schedules()
         self._register_listeners()
-        self._app.start(port=port)
+
+        app_handler = SlackRequestHandler(self._app)
+        api = FastAPI()
```
Comment (Germandrummer92) [inline sched_slack_bot/controller.py:97]: not completely against this, but somehow i think I would prefer starting FastAPI outside of the controller (e.g. in the bin file?), it's not related to the bot just to the kubernetes health check, right?

**Review by Germandrummer92 — CHANGES_REQUESTED** [review]

File: sched_slack_bot/controller.py (hunk 5 above)
Comment (kosta) [reply sched_slack_bot/controller.py:97]: Well the FastAPI object routes incoming http requests to the healthcheck _and_ to the bot. See the two usages of `api` annotations just below

Comment (Germandrummer92) [reply sched_slack_bot/controller.py:97]: yes i know, but we even added the full fastApi dependency just to do the extra health route. im also okay with it like this, but seems like a lot just to expose a health check route, and imho it just doesn't belong in the SchedSlackBotController (since it doesn't need a health check)
